### ISSUE TITLE
Add functionality to bring in phenotype information 

### ISSUE THREAD

**Issue by gregorgorjanc** [AlphaGenes/AlphaPeel#47]
Working with Minja Zorc, Mateja Janes, and Ros Craddock we will be adding support to bring phenotype information into AlphaPeel in addition to pedigree and (observed) genotype information.

Here are some theory notes

![IMG_3368](https://github.com/AlphaGenes/AlphaPeel/assets/725834/73851b29-6938-4c12-9271-15cb0fda57c7)

There are several actions here:

- [ ] Add functionality to read in phenotype file
- [ ] Add functionality to read in a user-defined phenotype penetrance function (we will start with a binary phenotype with recessive expression 
- [ ] Add functionality to use phenotype information for the phenotyped individuals
- [ ] Add functionality for imputing phenotype (inferring phenotype probability distribution)
- [ ] Probably some more actions!?

**Comment by gregorgorjanc** [comment]
Sharing chat/wall talk notes from today here - in defining these functions (penetrance function) we should always think from the perspective of a generative model, so `phenotype <- f(genotype)` and the `f()` is the penetrance function. This means (looking at the photo below) we should pick column 0 for the phenotype 0 and column 1 for phenotype 1 etc.

<img width="963" alt="Screenshot 2023-07-19 at 10 59 18" src="https://github.com/AlphaGenes/AlphaPeel/assets/725834/59e42813-fac4-4e33-a1c4-c1560c84204e">

**Comment by RosCraddock** [comment]
### Plan for the integration of phenotype information into AlphaPeel
We will start with monogenic recessive, single-locus.
**Inputs**
-pheno_file: Phenotype file with id, phenotype result, age of test (optional)
-pheno_penetrance_file: user-inputed matrix of the phenotype penetrance for p(observed phenotype | true genotype). Here, e can be inputted by the user, use default (e.g. 0.01, tbd), or estimated. An example input for this would be as follows:


1-e | e
-- | --
1-e | e
1-e | e
e | 1-e


where columns are Healthy and Diseased, and the rows represent the phased genotype states. Here, we will start with binary phenotype, but later explore inclusion of multi-categorical phenotype (e.g with coat colour). Where there is a difference due to the age of onset, potential for two phenotype penetrance files or an extended single phenotype penetrance file (i.e first two columns for Healthy and Diseased before onset, and second two after onset)

-pheno_error_input: user-inputted phenotype error rate (default is 0.01)
-age_of_onset: The age of clinical onset for phenotype (optional)
-est_pheno_error: flag to estimate the phenotype error
-pheno_probs: flag to write out the phenotype probabilities from P(true phenotype|true genotype)

**Outputs**
-pheno_error: Always write out if pheno_file is used.
-pheno_probs: Only printed if requested by user.

**Updates in Peeling**
Additional variables to collect in the pedigree and peelingInfo classes
Updates to the penetrance term
Function to calculate the phenotype probabilities (genoProbs*phenoPenetrance)
Function to estimate the phenotype error/penetrance

**Testing**
Functional Tests

- [ ] Take in the phenotype file with phenotype penetrance input and output pheno_error
- [ ] Take in the phenotype file with age of onset and phenotype penetrance input
- [ ] Use the user-input phenotype error
- [ ] Estimate the phenotype penetrance error
- [ ] Output phenotype probabilities with flag -pheno_probs
- [ ] Print error/warning message, stating phenotype_file ignored if phenotype penetrance file is not used.
- [ ] Print error/warning message stating "No age of onset so age at test ignored" if phenotype input includes age at test but no age_of_onset.

Accuracy Tests

- [ ] Phenotype file with penetrance and default error values
- [ ] Phenotype file with penetrance and estimated error values
- [ ] Phenotype file, including age of test, with age of onset and estimated error values

Future considerations: Using multiple phenotype traits and/or multiple loci.

[I will continue to update this comment]

@gregorgorjanc, please, may you add a branch for this modification: feat_pheno_Info?

**Comment by gregorgorjanc** [comment]
@RosCraddock I suggest you fork the devel branch and then you can easily create a new branch `feat_pheno_Info` on your end and once you will create a PR we will accept that branch on this end and eventually merge it into devel. I started doc work for this in https://github.com/AlphaGenes/AlphaPeel/commit/98f09825d6e6605671f6f23479ee84a52ed2b23e, but this branch is now way behind the devel, so I will delete it (there is nothing dramatic in there anyway!).

I suggest you don't worry about the age of onset just yet. Focus on the simplest case first and then we will see how to make it more general later. I am thinking here that we should maybe provide a way for the user to specify penetrance function somehow in  a general manner. The challenge here is that we will want this to be extendable to multiple loci ... But we can do this once we have the simple example implemented. One way also to deal with age of onset would be to define 4 phenotypes (0_early, 1_early, 0_late, 1_late) with corresponding penetrance function;)

**Comment by RosCraddock** [comment]
> ### Plan for the integration of phenotype information into AlphaPeel
> We will start with monogenic recessive, single-locus. **Inputs** -pheno_file: Phenotype file with id, phenotype result, age of test (optional) -pheno_penetrance_file: user-inputed matrix of the phenotype penetrance for p(observed phenotype | true genotype). Here, e can be inputted by the user, use default (e.g. 0.01, tbd), or estimated. An example input for this would be as follows:
> 
> 1-e	e
> 1-e	e
> 1-e	e
> e	1-e
> where columns are Healthy and Diseased, and the rows represent the phased genotype states. Here, we will start with binary phenotype, but later explore inclusion of multi-categorical phenotype (e.g with coat colour). Where there is a difference due to the age of onset, potential for two phenotype penetrance files or an extended single phenotype penetrance file (i.e first two columns for Healthy and Diseased before onset, and second two after onset)
> 
> -pheno_error_input: user-inputted phenotype error rate (default is 0.01) -age_of_onset: The age of clinical onset for phenotype (optional) -est_pheno_error: flag to estimate the phenotype error -pheno_probs: flag to write out the phenotype probabilities from P(true phenotype|true genotype)
> 
> **Outputs** -pheno_error: Always write out if pheno_file is used. -pheno_probs: Only printed if requested by user.
> 
> **Updates in Peeling** Additional variables to collect in the pedigree and peelingInfo classes Updates to the penetrance term Function to calculate the phenotype probabilities (genoProbs*phenoPenetrance) Function to estimate the phenotype error/penetrance
> 
> **Testing** Functional Tests
> 
> * [ ]  Take in the phenotype file with phenotype penetrance input and output pheno_error[ ]  Take in the phenotype file with age of onset and phenotype penetrance input[ ]  Use the user-input phenotype error[ ]  Estimate the phenotype penetrance error[ ]  Output phenotype probabilities with flag -pheno_probs[ ]  Print error/warning message, stating phenotype_file ignored if phenotype penetrance file is not used.[ ]  Print error/warning message stating "No age of onset so age at test ignored" if phenotype input includes age at test but no age_of_onset.
> 
> Accuracy Tests
> 
> * [ ]  Phenotype file with penetrance and default error values[ ]  Phenotype file with penetrance and estimated error values[ ]  Phenotype file, including age of test, with age of onset and estimated error values
> 
> Future considerations: Using multiple phenotype traits and/or multiple loci.
> 
> [I will continue to update this comment]
> 
> [@gregorgorjanc](https://github.com/gregorgorjanc), please, may you add a branch for this modification: feat_pheno_Info?

**Comment by RosCraddock** [comment]
**Updated Plan**

*Inputs*
-pheno_file: Phenotype file with id, phenotype result, age of test (optional).
-pheno_penetrance_file: user-inputed matrix of the phenotype penetrance.

*Peeling Arguments*
-update_pheno_penetrance: Flag to re-estimate the phenotype penetrance after each peeling cycle.

*Outputs*
-pheno_probs: flag to output phenotype probabilities from genotype probabilities and phenotype penetrance value
-pheno_penetrance: flag to output the phenotype penetrance

**Completed**
-Inclusion of phenotype data and phenotype penetrance. 
-Phenotype data can be multiple categories (not just binary), provided there is a phenotype penetrance inputted allows for this. i.e if phenotype data ranges from 0 to 4, then the phenotype penetrance input must have 5 columns.
-Can handle multiple phenotype records per individual.
But, only for a single locus genotype input and a single phenotype trait

**Still to do**
-update_pheno_penetrance
-mapping the inputted phenotype data to specific genotype locus (or loci)

**update_pheno_penetrance**
-Implemented Kinghorn (2003)'s described method for updating the phenotype penetrance after each peeling cycle.
-*However*, this method assumes high quality and quantity of phenotype data. Although the paper mentioned the updated penetrance values will be heavily conditioned by the initial penetrance values, this is not the case where it is limited and there is lower certainty in the genotype probability estimates. I observed this when testing small examples.
-So, I propose adding weightings based on the available phenotype data and/or certainty in the genotype probabilities. This would be:

newPhenoPenetrance = W*pheno_penetrance_input+(1-W)*updated_pheno_penetrance

*To do*
- Decide how to estimate the weighting
- Explore alternative methods - any suggestions?

**Mapping phenotype data to specific locus or loci**
This is a much later step that I have yet to finalise. I can see requiring another user input, and possible changes to the pheno_penetrance_file where we are accounting for epistatic effects.

**Comment by gregorgorjanc** [comment]
> This is a little vague and will likely evolve when I start to work on this final work package of my project. [@XingerTang](https://github.com/XingerTang) Given this, would it be preferable to change PR [#195](https://github.com/AlphaGenes/AlphaPeel/pull/195) into a "pheno_info" work-in-progress branch (as opposed to devel)?

I think that if a single locus functionality works and we throw an error if someone wants to pass in multiple phenotypes, then I think it’s ok to add into devel.

A quick test though. Say I want o run AlphaPeel with multiple loci and just one phenotype, how do I specify this? This is a legit case, say with SNP arrays that have one genetic disorder/marker and we also can observe phenotypes for that locus. Or does the current implementation work for a single locus only(which is ok for one milestone!) and we throw an error when a user passes in multiple loci but one phenotype?

The next two milestones would be to:
* Pass multiple loci and just one phenotype, but we need a way to link the one phenotype to a locus (hopefully in an extendable way for the next phase)
* Pass multiple loci and multiple phenotypes, but likely fewer phenotypes/traits than loci, hence the need for flexibility
* We could then think of more complex cases down the road.
@RosCraddock best if you open separate issues for these milestones, though it might make sense to meet and discuss how the ultimate solution might look like so we don’t paint ourselves into a corner.

**Comment by RosCraddock** [comment]
@gregorgorjanc Ok, I will add an error for now.

At the moment, I expect AlphaPeel will run, but the phenotype data will update all loci, rather than just the one of interest, since there is no way to specify this. As you say, this is the next step! I intended to throw an error for this in the meantime, but I can see I did not, so I will add this for PR #195.

Although for the scenario described, this would only require a new user input file: an incidence matrix of genotypes by phenotypes. Then, use this to identify the location of the locus or loci to update the genotype probabilities using the phenotype penetrance. If there are multiple loci but the phenotypic consequences are assumed independent, we may need multiple phenotype penetrance values (perhaps stored in an array/list). If there are multiple loci and the phenotype is considered polygenic, then we may need a singular phenotype penetrance that considers the different potential inheritance patterns and phenotypic expression (such as coat colour and epistasis). But, how do we specify this difference? 

To summarise, the cases could be:

- [X] One monogenic trait/disease with one phenotype measured and single locus genotype recorded
- [ ] One monogenic trait/disease with one phenotype measured, caused by a single locus capture in a multi-locus genotype
- [ ] One monogenic trait/disease with one phenotype measured, a multi-locus genotype with two or more independent causative loci (e.g PRA in the Golden Retriever has three different causative loci captured on the genetic testing panel).
- [ ] Multiple different monogenic traits/diseases, with some having a phenotype record, some not.
- [ ] Polygenic trait, such as coat colour under epistasis.
& many others

It would be good to discuss this. We have a meeting on Friday, so I will add it to the agenda.

**Comment by XingerTang** [comment]
This initial stage is already completed. Further development is going to AlphaGenes/AlphaPeel#197
